File: 강의 자료/02-알고리즘/07~08-두 포인터/문제별 코드/3273-두 수의 합/solution.cpp
#include <iostream>
#include <algorithm>
#include <string>
typedef long long int ll;
using namespace std;

#define NM 500005

int N, S;
int A[NM];

void input() {
	cin >> N;
	for (int i = 1; i <= N; i++) cin >> A[i];
	cin >> S;
}

void pro() {
	sort(A + 1, A + 1 + N);
	int ans = 0;
	for (int L = 1, R = N; L < R;){
		if (A[L] + A[R] == S) ans++;
		if (A[L] + A[R] >= S) R--;
		else L++;
	}
	cout << ans;
}


int main() {
	ios::sync_with_stdio(false); cin.tie(0); cout.tie(0);
	input();
	pro();
	return 0;
}

File: 강의 자료/02-알고리즘/07~08-두 포인터/문제별 코드/3273-두 수의 합/solution.py
import sys
si = sys.stdin.readline
n = int(si())
a = list(map(int, si().split()))
S = int(si())
a.sort()
L, R, ans = 0, n - 1, 0
while L < R:
    if a[L] + a[R] == S:
        ans += 1
    if a[L] + a[R] >= S:
        R -= 1
    else:
        L += 1
print(ans)